spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/AbstractSpannerRequestOptions.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.TimestampBound;
+
+/**
+ * Abstract class of common Read and Query request settings.
+ *
+ * @author Chengyuan Zhao
+ */
+public class AbstractSpannerRequestOptions<A, B extends AbstractSpannerRequestOptions> {
```

Comment:
I don't think you need this second class parameter `B extends AbstractSpannerRequestOptions`. The cast to `B` is not really safe anyway.
Why not just return `AbstractSpannerRequestOptions`?

Reply:
The `B` is for the inheriting children class to provide their own class type. The reason is we return `this` at the end of the setter methods so the user can chain them together. Without the `B` and returning just the abstract class the child classes' setters would return the parent abstract class on the final setter call so the following wouldn't compile because the final setter returns the parent abstract class type and not the child type: 

```
SpannerReadOptions = new SpannerReadOptions.setblah(xyz).setblahblah(abc).setblahblahblah(123);
```
The final setter call still just returns the abstract parent class.

Reply:
@meltsufin  the cast is safe here because B must extend the abstract parent class in the class definition and `this` is clearly of that same type.